Add Demo-2

abhi3700 opened the pull request:
## Description

This PR adds code + data for this logic:

Consider sample dataset of 1000 reviews.

1. For each sample in dataset (use the 1000 reviews)
    1. Create embedding of sample 
    2. Save embeddings to disk, linked with source sample
2. For each embedding in embeddings
    1. for each nbits = [8, 16, 32, 64, 128]
        1. Apply LSH to the embedding
        2. Add hash to a hash table, along with index of embedding
        3. Save hashes to disk (categorize by parameter)

Here is the [output file](https://github.com/abhi3700/semantic-hashing-demo/blob/demo-2/output/embeddings_w_lsh.csv) (in CSV format) with following columns:

| Text | Embedding | Hash 8-bit | Hash 16-bit | Hash 32-bit | Hash 64-bit |Hash 128-bit
|--|--|--|--|--|--|--|

abhi3700 committed:
Minor fix

Remove unused arg from func: 'hash_vector'

abhi3700 committed:
Change bucket to buckets

Actually there are multiple bucket with respective key hash. So, changed the name.

abhi3700 committed:
Add demo-2 step-1

- Replace <br /> with whitespace
- Add embedding corresponding to each review

abhi3700 committed:
Refator code

abhi3700 committed:
Add code for embeddings + LSH for each hyperplanes params

abhi3700 committed:
Changed the name

abhi3700 committed:
Update readme for preprocessing of data into embeddings

abhi3700 committed:
Output data file renamed, Add detection code to just read file

abhi3700 committed:
Minor changes

abhi3700 commented:
One observation when repeating data pre-processing, for most of the text reviews out of 1000, the values get changed after 3-4 decimal places. https://github.com/abhi3700/semantic-hashing-demo/commit/63d48d9de752e87d851f6499c65eaa86807be054
<img width="1007" alt="image" src="https://github.com/abhi3700/semantic-hashing-demo/assets/16472948/7b0eb356-c538-4f4e-8c19-11f7bb21dabc">

I don't think it would affect in bucketing.

abhi3700 commented:
This is the error at when calling function: pl.read_csv() directly…
```sh
polars.exceptions.ComputeError: could not parse `00011011110011110011001001110010` as dtype `i64` at column 'Hash 32-bit' (column number 5)

The current offset in the file is 34598 bytes.

You might want to try:
- increasing `infer_schema_length` (e.g. `infer_schema_length=10000`),
- specifying correct dtype with the `dtypes` argument
- setting `ignore_errors` to `True`,
- adding `00011011110011110011001001110010` to the `null_values` list.

Original error: ```remaining bytes non-empty```
```

Although this can be fixed by reading as `str` instead of default `i64` type. Not sure if there is any solution in vector DB or not 🤔  .
Now, it’s fixed. [commit](https://github.com/abhi3700/semantic-hashing-demo/blob/f1ddcb7e8dcdc2f0e45ac371b32e21ed43567633/src/semantic_hashing_demo/detection.py)

abhi3700 committed:
Add code + data for bucketing

abhi3700 committed:
Add code for detecting query text

abhi3700 committed:
Modify preprocessing with new design pattern of code

abhi3700 committed:
Update bucket to df

its-colby commented on src/semantic_hashing_demo/preprocessing.py:
```diff
@@ -0,0 +1,47 @@
+        lsh = LSH(nbits=nbits, seed=seed, embedding_size=embedding_size)
+
+        # Generate embeddings for each review
+        embeddings = [LSH.get_embedding(review, model) for review in reviews]
```
slight optimization: return np arrays from get_embedding, then set embeddings to the matrix, then do matrix mult for hash_vector instead of looping over every vector

abhi3700 replied:
Thanks for pointing this out. Fixed [here](https://github.com/abhi3700/semantic-hashing-demo/commit/b9bb4128eac75f4e0ae2f6fb6118d9d81a3f1b99).

I actually had used batching in the [rust counterpart](https://github.com/abhi3700/semantic-hashing-demo-rs/blob/cfaf1b37fd5812ba3d934ede6ab0e00a6955a2fe/src/utils.rs#L32). But, somehow forgot to add it here.

abhi3700 committed:
Revamp code design pattern for preprocessing, Also modify data structure

abhi3700 committed:
Revamp detection code with OOP design pattern

- fmt done
- `main.py` is removed now. As pre-processing & detection are run in separate scripts.
- 1st review as query text is checked. Observation 🔍 is that except for nbits=8, it falls into the expected bucket.

abhi3700 committed:
Optimize code

Refrain calling 'client.embeddings.create' multiple times in loop as it 'create' fn supports calling multiple text & return embeddings for all in matrix or n-dim array

abhi3700 committed:
Optimize code with re-runs

- Execution time is reduced from 15 mins to 30s for all hyperplanes.
- openai embedding 'create' fn supports list of texts. So, loop removed from main() fn

abhi3700 committed:
Buckets hash changes when re-run with same seed

- For 8 nbits, some buckets modify with less/more texts falling inside.
- For other nbits: 16, 32, 64, 128, some buckets change their hash with same texts inside.
 This is the observation seed when re-running again. In this commit, the changes can be seen.

abhi3700 committed:
Buckets hash changes when re-run again with same seed

- For 8 nbits, some buckets modify with less/more texts falling inside. Also, their hash changes as well.
- For other nbits: 16, 32, 64, 128, some buckets change their hash with same texts inside.
 This is the observation seed when re-running again. In this commit, the changes can be seen. It seems like 8 nbits is not reliable.

abhi3700 committed:
Update detect text code with optimized LSH class

abhi3700 committed:
Restore preprocessed_data file generation

Modified the script to include generating preprocessed_data with embedding i.e. 'List[float]' for each review along with buckets generation for all hyperplanes

abhi3700 committed:
Add @jeremy code for generating text

abhi3700 committed:
Refactor code

Fix static typing. Added pre-check before detecting text query.

abhi3700 committed:
Add code for processing generated paragraphs

abhi3700 committed:
Add code for saving Hamming Distance (HD) matrix

abhi3700 commented:
Observation (for 20 paragraphs): The matrix does have the relatively lowest Hamming distance diagonally as expected. You can find those in `matrix_8.csv`, `matrix_16.csv`, etc.

abhi3700 committed:
Add README with clean instruction for setup, Fmt done

abhi3700 committed:
Add interactive heatmap plot (in html) for the matrices for each nbits

abhi3700 committed:
Include build inside install step itself

abhi3700 commented:
Added interactive (in html format…..hover on it for values) heatmap plots to give a visualization of the generated matrices for each nbits = 8, 16, 32, 64, 128 for 20 paragraphs b/w original & slightly modified texts. Download the html plots from [commit](https://github.com/abhi3700/semantic-hashing-demo/commit/379910cf060f4c0e6a8886fbfb842029e16aa7d1).
:eyes:  With increasing no. of hyperplanes, the diagonal blue line is more crystal clear.

| ![image](https://github.com/abhi3700/semantic-hashing-demo/assets/16472948/475f3412-a138-4784-92ee-b2531b9aea6a) | ![image](https://github.com/abhi3700/semantic-hashing-demo/assets/16472948/c0ee5a64-c1ba-4be4-a02f-73302b6a88ec)  | 
|--|--|

![image](https://github.com/abhi3700/semantic-hashing-demo/assets/16472948/58f44bc5-7d74-4ba1-a86c-7ef6122abf1f)

| ![image](https://github.com/abhi3700/semantic-hashing-demo/assets/16472948/28cea33b-8191-48f4-8c83-b37986afb26c) | ![image](https://github.com/abhi3700/semantic-hashing-demo/assets/16472948/7717b6d5-7d29-4aa8-9a19-d06fe2d401cf) |
|--|--|

abhi3700 committed:
Add code for generating plots, Remove dataframe alongwith matrix_nbits.csv files

- Add 3 more python packages for this to requirements.txt, pyproject.toml files

0xjeremyfrank commented on src/semantic_hashing_demo/generate_data.py:
```diff
@@ -0,0 +1,60 @@
+    # semantically_similar:str
+
+
+# data = []
+# for _ in range(5):
+#     data.extend(marvin.generate(n=3, target=ParagraphData, instructions="generate paragraphs for comparison testing. the paragraphs should be almost identical, with only a few words changed. each paragraph should be at least 100 words long."))
+#     print(data)
+# data_dicts = [d.dict() for d in data]
+# df = pl.DataFrame(data_dicts)
+# df.write_csv("./data/paragraphs.csv")
```
This can be deleted.

abhi3700 replied:
[Commit](https://github.com/abhi3700/semantic-hashing-demo/pull/2/commits/506a8c9ea571e9706b10a1b80b2ac0ff8959dd61).

0xjeremyfrank commented on src/semantic_hashing_demo/lsh.py:
```diff
@@ -0,0 +1,156 @@
+#     # query = "I have bought many of the Vitality canned dog food products and have found them all to be of good quality. The product looks more like a stew than a processed meat and it smells good. My Labrador is finicky and she likes this product better than  most."  # changed the 1st review a bit
+#     # query = 'Product reached marked as Jumbo Salted Peanuts...the peanuts were actually small sized unsalted. Not sure if this was a mistake or if the vendor wanted to indicate the product as "Jumbo".'  # changed the 2nd review a bit
+#     # query = 'Great taffy at a better price.  There was a broad assortment of yummy taffy.  Delivery was super fast.  If your a taffy lover, this is a good chance.'  # changed the 5th review a bit
+
+#     # ===== Type-3 =====
+#     # query = "I've purchased numerous cans of the Vitality dog food line and have consistently found them to be of high quality. They resemble stew more than they do processed meat, and they have a more pleasant aroma. My picky Labrador prefers this brand over many others." # ai-generated the 1st review a bit
+#     # query = 'Excellent value for delicious taffy. The selection offered a broad variety of delectable flavors. The shipping was impressively fast. For enthusiasts of taffy, this offer is a must-grab.'    # ai-generated the 5th review a bit
+#     # query = "This dog food is highly nutritious and beneficial for digestive health. It's also suitable for young puppies. My dog consistently consumes the recommended portion at each meal."  # ai-generated the index-9 review a bit
+
+#     # ===== Type-4 =====
+#     # query = "I have lived out of the US for over 7 yrs now, and I so miss my Twizzlers!!  When I go back to visit or someone visits me, I always stock up.  All I can say is YUM!<br />Sell these in Mexico and you will have a faithful buyer, more often than I'm able to buy them right now."
+#     # query = "Good oatmeal.  I like the apple cinnamon the best.  Though I wouldn't follow the directions on the package since it always comes out too soupy for my taste.  That could just be me since I like my oatmeal really thick to add some milk on top of."
+#     # query = "I roast at home with a stove-top popcorn popper (but I do it outside, of course). These beans (Coffee Bean Direct Green Mexican Altura) seem to be well-suited for this method. The first and second cracks are distinct, and I've roasted the beans from medium to slightly dark with great results every time. The aroma is strong and persistent. The taste is smooth, velvety, yet lively. ...
+#     hash_query = hash_vector(get_embedding(query), plane_norms)
+#     # hash_query = hash_vector(get_embedding(query, model), plane_norms)
+#     print(f"\nFor a given text: \"{query}\", it's computed hash is '{hash_query}'.")
+
+#     # calculate the hamming distance between the query and each bucket
+#     print("\nhamming distances b/w the query from each bucket key:")
+#     hamming_distances = []
+#     for hash_str in buckets.keys():
+#         hamming_distances.append(hamming_distance(hash_query, hash_str))
+#     print(hamming_distances)
+#     if contains_zero(hamming_distances):
+#         print(
+#             "🙂 The given text falls into the bucket with its key having exact same hash"
+#         )
+#     else:
+#         print(
+#             "☹️ As no exact hash found, deliberately the closest bucket with min. hamming distance is selected here from left --> right."
+#         )
+#     # Get the index of the lowest one
+#     min_index = np.argmin(hamming_distances)
+#     print(
+#         f"\nHence, the given text belongs to the index-{min_index} of bucket list, \ni.e. the bucket with key: '{list(buckets.keys())[min_index]}', value: [{list(buckets.values())[min_index]}]."
+#     )
+
+
+# if __name__ == "__main__":
+#     main()
```
This can be deleted also.

abhi3700 replied:
[Commit](https://github.com/abhi3700/semantic-hashing-demo/pull/2/commits/506a8c9ea571e9706b10a1b80b2ac0ff8959dd61).

abhi3700 committed:
Cleanup
